# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    print("-------------------------------           Source Text              ---------------------------------------\n")
    print(source_text)
    print("\n----------------------------------------End Of Source Text------------------------------------------------")
    
    print("\n\n")
    
    print("-------------------------------          Destination Text          ---------------------------------------\n")
    print(source_text)
    print("\n----------------------------------------End Of Destination Text -------------------------------------------")

    
    print("\n\n")
    source_id_text = [[source_vocab_to_int.get(word, source_vocab_to_int['<UNK>']) for word in line.split()] for line in source_text.split("\n")]
    target_id_text = [[target_vocab_to_int.get(word, target_vocab_to_int['<UNK>']) for word in line.split()] + [target_vocab_to_int['<EOS>']] for line in target_text.split("\n")]

    
    return (source_id_text, target_id_text)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

-------------------------------           Source Text              ---------------------------------------

new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .

----------------------------------------End Of Source Text------------------------------------------------



-------------------------------          Destination Text          ---------------------------------------

new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .

----------------------------------

### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.


# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    inputs = tf.placeholder(tf.int32,[None,None], name="input")
    targets = tf.placeholder(tf.int32, [None,None], name="targets")
    learning_rate = tf.placeholder(tf.float32, shape=[],  name="learning_rate")
    keep_prob = tf.placeholder(tf.float32, shape=[], name="keep_prob")
    target_sequence_length = tf.placeholder(tf.int32, (None,), name="target_sequence_length")
    max_target_sequence_length = tf.reduce_max(target_sequence_length, name="max_target_len")
    source_sequence_length = tf.placeholder(tf.int32, (None,), name="source_sequence_length")
    
    
    
    return (inputs, targets, learning_rate, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [8]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    ending = tf.strided_slice(target_data, [0,0], [batch_size,-1], [1,1])
    dec_input = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)
    return dec_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [9]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    encoding_embedding_sequence = tf.contrib.layers.embed_sequence(rnn_inputs, source_vocab_size, encoding_embedding_size)
    def build_cell(rnn_size):
        cell = tf.contrib.rnn.LSTMCell(rnn_size, initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        return cell
    
    multi_LSTMCell = tf.contrib.rnn.MultiRNNCell([build_cell(rnn_size) for _ in range (num_layers)])
    
    encoder_output, encoder_state = tf.nn.dynamic_rnn(multi_LSTMCell, encoding_embedding_sequence, sequence_length=source_sequence_length, dtype=tf.float32)
    
    return (encoder_output, encoder_state)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [10]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    training_helper = tf.contrib.seq2seq.TrainingHelper(inputs=dec_embed_input, sequence_length=target_sequence_length, 
                                                        time_major=False)
    training_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, training_helper, encoder_state, output_layer)

    training_decoder_output = tf.contrib.seq2seq.dynamic_decode(training_decoder, impute_finished=True,
                                                                maximum_iterations=max_summary_length)[0]
    #training_decoder_ouput = tf.nn.dropout(training_decoder_output, keep_prob=keep_prob)
    
    
    return training_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [11]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function

    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name='start_tokens')
    #start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name='start_tokens')



    infer_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings,start_tokens, end_of_sequence_id )
    infer_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, infer_helper, encoder_state, output_layer)
    infer_decoder_output = tf.contrib.seq2seq.dynamic_decode(infer_decoder, impute_finished=True,
                                                            maximum_iterations=max_target_sequence_length)[0]

         

    
    return infer_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [12]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    
    start_of_sequence_id = target_vocab_to_int['<GO>']
    end_of_sequence_id = target_vocab_to_int['<EOS>']


    decoder_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(decoder_embeddings, dec_input)
    
    def build_dec_cell(rnn_size):
        dec_cell = tf.contrib.rnn.LSTMCell(rnn_size, initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        return dec_cell
    
    dec_cell = tf.contrib.rnn.MultiRNNCell([build_dec_cell(rnn_size) for i in range (num_layers)])
    
    
    output_layer = Dense(target_vocab_size, kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))
    
    with tf.variable_scope("decode"):
        
        train_decoder_output = decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length,
                                               max_target_sequence_length, output_layer, keep_prob)
        
    with tf.variable_scope("decode", reuse=True):
    
        infer_decoder_output = decoding_layer_infer(encoder_state, dec_cell, decoder_embeddings, start_of_sequence_id, 
                                                 end_of_sequence_id, max_target_sequence_length, target_vocab_size,
                                                     output_layer, batch_size, keep_prob)
    
    return (train_decoder_output, infer_decoder_output)



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [13]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    
    enc_output, enc_state = encoding_layer(input_data, rnn_size, num_layers, keep_prob,  
                   source_sequence_length, source_vocab_size, enc_embedding_size)
    dec_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)
    train_decoder_output, infer_decoder_output = decoding_layer(dec_input, enc_state, target_sequence_length, 
                                                                max_target_sentence_length, rnn_size, num_layers, 
                                                                target_vocab_to_int, target_vocab_size, batch_size, 
                                                                keep_prob, dec_embedding_size)
    
    return (train_decoder_output, infer_decoder_output)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [14]:
# Number of Epochs
epochs = 5
# Batch Size
batch_size = 64
# RNN Size
rnn_size = 128
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 15
decoding_embedding_size = 15
# Learning Rate
learning_rate = 0.01
# Dropout Keep Probability
keep_probability = 0.5
display_step = 10

### Build the Graph
Build the graph using the neural network you implemented.

In [15]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [16]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [17]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch   10/2154 - Train Accuracy: 0.2906, Validation Accuracy: 0.3516, Loss: 3.3211
Epoch   0 Batch   20/2154 - Train Accuracy: 0.3523, Validation Accuracy: 0.3828, Loss: 2.9674
Epoch   0 Batch   30/2154 - Train Accuracy: 0.3633, Validation Accuracy: 0.4148, Loss: 2.7705
Epoch   0 Batch   40/2154 - Train Accuracy: 0.3668, Validation Accuracy: 0.4304, Loss: 2.6335
Epoch   0 Batch   50/2154 - Train Accuracy: 0.4030, Validation Accuracy: 0.4702, Loss: 2.4131
Epoch   0 Batch   60/2154 - Train Accuracy: 0.4125, Validation Accuracy: 0.4496, Loss: 2.2147
Epoch   0 Batch   70/2154 - Train Accuracy: 0.4194, Validation Accuracy: 0.4517, Loss: 2.0908
Epoch   0 Batch   80/2154 - Train Accuracy: 0.3797, Validation Accuracy: 0.4006, Loss: 1.9037
Epoch   0 Batch   90/2154 - Train Accuracy: 0.3783, Validation Accuracy: 0.4467, Loss: 1.7891
Epoch   0 Batch  100/2154 - Train Accuracy: 0.3966, Validation Accuracy: 0.4673, Loss: 1.6636
Epoch   0 Batch  110/2154 - Train Accuracy: 0.3805, Validati

Epoch   0 Batch  890/2154 - Train Accuracy: 0.7245, Validation Accuracy: 0.6733, Loss: 0.3180
Epoch   0 Batch  900/2154 - Train Accuracy: 0.6883, Validation Accuracy: 0.6463, Loss: 0.3540
Epoch   0 Batch  910/2154 - Train Accuracy: 0.7368, Validation Accuracy: 0.7124, Loss: 0.3119
Epoch   0 Batch  920/2154 - Train Accuracy: 0.7314, Validation Accuracy: 0.6818, Loss: 0.2707
Epoch   0 Batch  930/2154 - Train Accuracy: 0.7438, Validation Accuracy: 0.7301, Loss: 0.3015
Epoch   0 Batch  940/2154 - Train Accuracy: 0.7523, Validation Accuracy: 0.7010, Loss: 0.2961
Epoch   0 Batch  950/2154 - Train Accuracy: 0.7163, Validation Accuracy: 0.7266, Loss: 0.3095
Epoch   0 Batch  960/2154 - Train Accuracy: 0.7773, Validation Accuracy: 0.7109, Loss: 0.2857
Epoch   0 Batch  970/2154 - Train Accuracy: 0.8031, Validation Accuracy: 0.7280, Loss: 0.2765
Epoch   0 Batch  980/2154 - Train Accuracy: 0.7688, Validation Accuracy: 0.7045, Loss: 0.2554
Epoch   0 Batch  990/2154 - Train Accuracy: 0.7531, Validati

Epoch   0 Batch 1770/2154 - Train Accuracy: 0.8547, Validation Accuracy: 0.8658, Loss: 0.1139
Epoch   0 Batch 1780/2154 - Train Accuracy: 0.9117, Validation Accuracy: 0.7891, Loss: 0.0914
Epoch   0 Batch 1790/2154 - Train Accuracy: 0.9115, Validation Accuracy: 0.8565, Loss: 0.1033
Epoch   0 Batch 1800/2154 - Train Accuracy: 0.9195, Validation Accuracy: 0.8217, Loss: 0.1192
Epoch   0 Batch 1810/2154 - Train Accuracy: 0.8625, Validation Accuracy: 0.8267, Loss: 0.1119
Epoch   0 Batch 1820/2154 - Train Accuracy: 0.8938, Validation Accuracy: 0.8629, Loss: 0.0995
Epoch   0 Batch 1830/2154 - Train Accuracy: 0.8730, Validation Accuracy: 0.8786, Loss: 0.1689
Epoch   0 Batch 1840/2154 - Train Accuracy: 0.9340, Validation Accuracy: 0.8430, Loss: 0.0925
Epoch   0 Batch 1850/2154 - Train Accuracy: 0.9243, Validation Accuracy: 0.8260, Loss: 0.1503
Epoch   0 Batch 1860/2154 - Train Accuracy: 0.8997, Validation Accuracy: 0.8210, Loss: 0.0989
Epoch   0 Batch 1870/2154 - Train Accuracy: 0.8789, Validati

Epoch   1 Batch  500/2154 - Train Accuracy: 0.9490, Validation Accuracy: 0.9077, Loss: 0.0780
Epoch   1 Batch  510/2154 - Train Accuracy: 0.8773, Validation Accuracy: 0.9098, Loss: 0.0669
Epoch   1 Batch  520/2154 - Train Accuracy: 0.9227, Validation Accuracy: 0.9254, Loss: 0.0485
Epoch   1 Batch  530/2154 - Train Accuracy: 0.9437, Validation Accuracy: 0.9212, Loss: 0.0502
Epoch   1 Batch  540/2154 - Train Accuracy: 0.9202, Validation Accuracy: 0.9119, Loss: 0.0532
Epoch   1 Batch  550/2154 - Train Accuracy: 0.9256, Validation Accuracy: 0.9034, Loss: 0.0629
Epoch   1 Batch  560/2154 - Train Accuracy: 0.9289, Validation Accuracy: 0.8999, Loss: 0.0674
Epoch   1 Batch  570/2154 - Train Accuracy: 0.9516, Validation Accuracy: 0.9062, Loss: 0.0622
Epoch   1 Batch  580/2154 - Train Accuracy: 0.9441, Validation Accuracy: 0.9070, Loss: 0.0664
Epoch   1 Batch  590/2154 - Train Accuracy: 0.9013, Validation Accuracy: 0.9197, Loss: 0.0638
Epoch   1 Batch  600/2154 - Train Accuracy: 0.9465, Validati

Epoch   1 Batch 1380/2154 - Train Accuracy: 0.9570, Validation Accuracy: 0.9048, Loss: 0.0480
Epoch   1 Batch 1390/2154 - Train Accuracy: 0.8981, Validation Accuracy: 0.8892, Loss: 0.0586
Epoch   1 Batch 1400/2154 - Train Accuracy: 0.9161, Validation Accuracy: 0.9276, Loss: 0.0471
Epoch   1 Batch 1410/2154 - Train Accuracy: 0.9523, Validation Accuracy: 0.9219, Loss: 0.0588
Epoch   1 Batch 1420/2154 - Train Accuracy: 0.9482, Validation Accuracy: 0.8935, Loss: 0.0456
Epoch   1 Batch 1430/2154 - Train Accuracy: 0.9144, Validation Accuracy: 0.8871, Loss: 0.0591
Epoch   1 Batch 1440/2154 - Train Accuracy: 0.9375, Validation Accuracy: 0.9027, Loss: 0.0708
Epoch   1 Batch 1450/2154 - Train Accuracy: 0.8849, Validation Accuracy: 0.9318, Loss: 0.0659
Epoch   1 Batch 1460/2154 - Train Accuracy: 0.9102, Validation Accuracy: 0.9268, Loss: 0.0762
Epoch   1 Batch 1470/2154 - Train Accuracy: 0.9350, Validation Accuracy: 0.9375, Loss: 0.0549
Epoch   1 Batch 1480/2154 - Train Accuracy: 0.9172, Validati

Epoch   2 Batch  110/2154 - Train Accuracy: 0.9203, Validation Accuracy: 0.9496, Loss: 0.0755
Epoch   2 Batch  120/2154 - Train Accuracy: 0.9548, Validation Accuracy: 0.9070, Loss: 0.0441
Epoch   2 Batch  130/2154 - Train Accuracy: 0.9703, Validation Accuracy: 0.9254, Loss: 0.0310
Epoch   2 Batch  140/2154 - Train Accuracy: 0.9398, Validation Accuracy: 0.8991, Loss: 0.0694
Epoch   2 Batch  150/2154 - Train Accuracy: 0.9494, Validation Accuracy: 0.9347, Loss: 0.0415
Epoch   2 Batch  160/2154 - Train Accuracy: 0.9367, Validation Accuracy: 0.9375, Loss: 0.0514
Epoch   2 Batch  170/2154 - Train Accuracy: 0.9680, Validation Accuracy: 0.9624, Loss: 0.0325
Epoch   2 Batch  180/2154 - Train Accuracy: 0.9172, Validation Accuracy: 0.9616, Loss: 0.0450
Epoch   2 Batch  190/2154 - Train Accuracy: 0.9453, Validation Accuracy: 0.9652, Loss: 0.0363
Epoch   2 Batch  200/2154 - Train Accuracy: 0.9469, Validation Accuracy: 0.9510, Loss: 0.0296
Epoch   2 Batch  210/2154 - Train Accuracy: 0.9786, Validati

Epoch   2 Batch  990/2154 - Train Accuracy: 0.9437, Validation Accuracy: 0.9297, Loss: 0.0334
Epoch   2 Batch 1000/2154 - Train Accuracy: 0.9107, Validation Accuracy: 0.9247, Loss: 0.0384
Epoch   2 Batch 1010/2154 - Train Accuracy: 0.9586, Validation Accuracy: 0.9766, Loss: 0.0354
Epoch   2 Batch 1020/2154 - Train Accuracy: 0.9547, Validation Accuracy: 0.9418, Loss: 0.0517
Epoch   2 Batch 1030/2154 - Train Accuracy: 0.9252, Validation Accuracy: 0.9652, Loss: 0.0470
Epoch   2 Batch 1040/2154 - Train Accuracy: 0.9641, Validation Accuracy: 0.9609, Loss: 0.0297
Epoch   2 Batch 1050/2154 - Train Accuracy: 0.9266, Validation Accuracy: 0.9595, Loss: 0.0684
Epoch   2 Batch 1060/2154 - Train Accuracy: 0.9242, Validation Accuracy: 0.9439, Loss: 0.0404
Epoch   2 Batch 1070/2154 - Train Accuracy: 0.8973, Validation Accuracy: 0.9169, Loss: 0.0661
Epoch   2 Batch 1080/2154 - Train Accuracy: 0.9297, Validation Accuracy: 0.9673, Loss: 0.0556
Epoch   2 Batch 1090/2154 - Train Accuracy: 0.9655, Validati

Epoch   2 Batch 1870/2154 - Train Accuracy: 0.9609, Validation Accuracy: 0.9602, Loss: 0.0283
Epoch   2 Batch 1880/2154 - Train Accuracy: 0.9445, Validation Accuracy: 0.9396, Loss: 0.0509
Epoch   2 Batch 1890/2154 - Train Accuracy: 0.9740, Validation Accuracy: 0.9560, Loss: 0.0353
Epoch   2 Batch 1900/2154 - Train Accuracy: 0.9727, Validation Accuracy: 0.9538, Loss: 0.0336
Epoch   2 Batch 1910/2154 - Train Accuracy: 0.9868, Validation Accuracy: 0.9183, Loss: 0.0262
Epoch   2 Batch 1920/2154 - Train Accuracy: 0.9704, Validation Accuracy: 0.9439, Loss: 0.0309
Epoch   2 Batch 1930/2154 - Train Accuracy: 0.9508, Validation Accuracy: 0.9503, Loss: 0.0473
Epoch   2 Batch 1940/2154 - Train Accuracy: 0.9204, Validation Accuracy: 0.9688, Loss: 0.0433
Epoch   2 Batch 1950/2154 - Train Accuracy: 0.9898, Validation Accuracy: 0.9709, Loss: 0.0278
Epoch   2 Batch 1960/2154 - Train Accuracy: 0.9622, Validation Accuracy: 0.9489, Loss: 0.0397
Epoch   2 Batch 1970/2154 - Train Accuracy: 0.9406, Validati

Epoch   3 Batch  600/2154 - Train Accuracy: 0.9433, Validation Accuracy: 0.9467, Loss: 0.0451
Epoch   3 Batch  610/2154 - Train Accuracy: 0.9501, Validation Accuracy: 0.9254, Loss: 0.0428
Epoch   3 Batch  620/2154 - Train Accuracy: 0.9330, Validation Accuracy: 0.9205, Loss: 0.0387
Epoch   3 Batch  630/2154 - Train Accuracy: 0.9688, Validation Accuracy: 0.9496, Loss: 0.0335
Epoch   3 Batch  640/2154 - Train Accuracy: 0.9742, Validation Accuracy: 0.9261, Loss: 0.0274
Epoch   3 Batch  650/2154 - Train Accuracy: 0.9383, Validation Accuracy: 0.9219, Loss: 0.0440
Epoch   3 Batch  660/2154 - Train Accuracy: 0.9516, Validation Accuracy: 0.9403, Loss: 0.0377
Epoch   3 Batch  670/2154 - Train Accuracy: 0.9811, Validation Accuracy: 0.9418, Loss: 0.0232
Epoch   3 Batch  680/2154 - Train Accuracy: 0.9622, Validation Accuracy: 0.9403, Loss: 0.0305
Epoch   3 Batch  690/2154 - Train Accuracy: 0.9328, Validation Accuracy: 0.9126, Loss: 0.0544
Epoch   3 Batch  700/2154 - Train Accuracy: 0.9416, Validati

Epoch   3 Batch 1480/2154 - Train Accuracy: 0.9555, Validation Accuracy: 0.9545, Loss: 0.0336
Epoch   3 Batch 1490/2154 - Train Accuracy: 0.9453, Validation Accuracy: 0.9787, Loss: 0.0295
Epoch   3 Batch 1500/2154 - Train Accuracy: 0.9696, Validation Accuracy: 0.9602, Loss: 0.0340
Epoch   3 Batch 1510/2154 - Train Accuracy: 0.9507, Validation Accuracy: 0.9446, Loss: 0.0361
Epoch   3 Batch 1520/2154 - Train Accuracy: 0.9688, Validation Accuracy: 0.9858, Loss: 0.0335
Epoch   3 Batch 1530/2154 - Train Accuracy: 0.9597, Validation Accuracy: 0.9595, Loss: 0.0322
Epoch   3 Batch 1540/2154 - Train Accuracy: 0.9617, Validation Accuracy: 0.9531, Loss: 0.0439
Epoch   3 Batch 1550/2154 - Train Accuracy: 0.9750, Validation Accuracy: 0.9361, Loss: 0.0402
Epoch   3 Batch 1560/2154 - Train Accuracy: 0.9516, Validation Accuracy: 0.9737, Loss: 0.0330
Epoch   3 Batch 1570/2154 - Train Accuracy: 0.9453, Validation Accuracy: 0.9531, Loss: 0.0296
Epoch   3 Batch 1580/2154 - Train Accuracy: 0.9638, Validati

Epoch   4 Batch  210/2154 - Train Accuracy: 0.9836, Validation Accuracy: 0.9297, Loss: 0.0267
Epoch   4 Batch  220/2154 - Train Accuracy: 0.9688, Validation Accuracy: 0.9773, Loss: 0.0342
Epoch   4 Batch  230/2154 - Train Accuracy: 0.9375, Validation Accuracy: 0.9638, Loss: 0.0280
Epoch   4 Batch  240/2154 - Train Accuracy: 0.9539, Validation Accuracy: 0.9680, Loss: 0.0431
Epoch   4 Batch  250/2154 - Train Accuracy: 0.9664, Validation Accuracy: 0.9411, Loss: 0.0363
Epoch   4 Batch  260/2154 - Train Accuracy: 0.9609, Validation Accuracy: 0.9226, Loss: 0.0329
Epoch   4 Batch  270/2154 - Train Accuracy: 0.9762, Validation Accuracy: 0.9411, Loss: 0.0341
Epoch   4 Batch  280/2154 - Train Accuracy: 0.9326, Validation Accuracy: 0.9446, Loss: 0.0467
Epoch   4 Batch  290/2154 - Train Accuracy: 0.9194, Validation Accuracy: 0.9368, Loss: 0.0531
Epoch   4 Batch  300/2154 - Train Accuracy: 0.9285, Validation Accuracy: 0.9588, Loss: 0.0366
Epoch   4 Batch  310/2154 - Train Accuracy: 0.9794, Validati

Epoch   4 Batch 1090/2154 - Train Accuracy: 0.9531, Validation Accuracy: 0.9382, Loss: 0.0241
Epoch   4 Batch 1100/2154 - Train Accuracy: 0.9578, Validation Accuracy: 0.9616, Loss: 0.0366
Epoch   4 Batch 1110/2154 - Train Accuracy: 0.9688, Validation Accuracy: 0.9439, Loss: 0.0263
Epoch   4 Batch 1120/2154 - Train Accuracy: 0.9852, Validation Accuracy: 0.9304, Loss: 0.0310
Epoch   4 Batch 1130/2154 - Train Accuracy: 0.9375, Validation Accuracy: 0.9489, Loss: 0.0201
Epoch   4 Batch 1140/2154 - Train Accuracy: 0.9273, Validation Accuracy: 0.9290, Loss: 0.0356
Epoch   4 Batch 1150/2154 - Train Accuracy: 0.9589, Validation Accuracy: 0.9261, Loss: 0.0343
Epoch   4 Batch 1160/2154 - Train Accuracy: 0.9539, Validation Accuracy: 0.9247, Loss: 0.0312
Epoch   4 Batch 1170/2154 - Train Accuracy: 0.9680, Validation Accuracy: 0.9283, Loss: 0.0216
Epoch   4 Batch 1180/2154 - Train Accuracy: 0.9712, Validation Accuracy: 0.9794, Loss: 0.0254
Epoch   4 Batch 1190/2154 - Train Accuracy: 0.9594, Validati

Epoch   4 Batch 1970/2154 - Train Accuracy: 0.9336, Validation Accuracy: 0.9460, Loss: 0.0467
Epoch   4 Batch 1980/2154 - Train Accuracy: 0.9422, Validation Accuracy: 0.9411, Loss: 0.0313
Epoch   4 Batch 1990/2154 - Train Accuracy: 0.9539, Validation Accuracy: 0.9439, Loss: 0.0240
Epoch   4 Batch 2000/2154 - Train Accuracy: 0.9367, Validation Accuracy: 0.9453, Loss: 0.0496
Epoch   4 Batch 2010/2154 - Train Accuracy: 0.9539, Validation Accuracy: 0.9418, Loss: 0.0705
Epoch   4 Batch 2020/2154 - Train Accuracy: 0.9893, Validation Accuracy: 0.9524, Loss: 0.0145
Epoch   4 Batch 2030/2154 - Train Accuracy: 0.9516, Validation Accuracy: 0.9354, Loss: 0.0270
Epoch   4 Batch 2040/2154 - Train Accuracy: 0.9688, Validation Accuracy: 0.9652, Loss: 0.0507
Epoch   4 Batch 2050/2154 - Train Accuracy: 0.9227, Validation Accuracy: 0.9567, Loss: 0.0536
Epoch   4 Batch 2060/2154 - Train Accuracy: 0.9581, Validation Accuracy: 0.9553, Loss: 0.0425
Epoch   4 Batch 2070/2154 - Train Accuracy: 0.9531, Validati

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [18]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [19]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [20]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    #sentence_lower = sentence.lower()
    
    return [ vocab_to_int.get(word, vocab_to_int.get('<UNK>')) for word in sentence.lower().split() ]


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [21]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [72, 114, 225, 62, 35, 197, 8]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [73, 356, 196, 176, 178, 62, 4, 5, 1]
  French Words: il a vu un vieux camion noir . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.